src/main/groovy/pl/allegro/tech/build/axion/release/domain/Releaser.groovy

```diff
@@ -35,12 +37,33 @@ class Releaser {
             repository.tag(tagName)
 
             hooksRunner.runPostReleaseHooks(rules.hooks, rules, versionContext, version)
+            return Optional.of(tagName)
         } else {
             logger.quiet("Working on released version ${version}, nothing to release.")
+            return Optional.empty()
+        }
+    }
+
+    void releaseAndPush(Properties rules) {
+        def releasedTagName = release(rules)
+
+        try {
+            pushRelease()
+        } catch (ScmPushFailedException | GitAPIException e) {
+            logger.quiet("Exception occurred during pushing the release: ${e.message}.")
+            releasedTagName.ifPresent{ tagName ->
+                logger.quiet("Removing tag: $tagName.")
+                rollbackRelease(tagName)
```

Comment:
i would add "tag $tagName removed" and wrap rollbackRelease in try/catch, so if any error occurs in between, we know that the tag has not been removed